homeassistant/components/recorder/purge.py

```diff
@@ -101,3 +125,193 @@ def purge_old_data(instance, purge_days: int, repack: bool) -> bool:
+
+
+def _purge_filtered_data(
+    instance: Recorder,
+    session: Session,
+    *,
+    batch_size_hours: int = 1,
+) -> bool:
+    """Purge filtered states and events that shouldn't be in database."""
+
+    _LOGGER.debug("Purging filtered states and events")
+    utc_now = dt_util.utcnow()
+    batch_purge_before_states = utc_now
+    batch_purge_before_events = utc_now
+
+    # Check if excluded entity_ids are in database
+    excluded_entity_ids: List[str] = [
+        entity_id
+        for (entity_id,) in session.query(distinct(States.entity_id)).all()
+        if not instance.entity_filter(entity_id)
+    ]
+
+    if len(excluded_entity_ids) != 0:
+        batch_purge_before_states = _purge_filtered_states(
+            session,
+            excluded_entity_ids,
+            batch_purge_before_states,
+            batch_size_hours,
+        )
+
+    # Check if excluded event_types are in database
+    excluded_event_types: List[str] = [
+        event_type
+        for (event_type,) in session.query(distinct(Events.event_type)).all()
+        if event_type in instance.exclude_t
+    ]
+
+    if len(excluded_event_types) != 0:
+        batch_purge_before_events = _purge_filtered_events(
+            session,
+            excluded_event_types,
+            batch_purge_before_events,
+            batch_size_hours,
+        )
+
+    if batch_purge_before_states < utc_now or batch_purge_before_events < utc_now:
+        _LOGGER.debug("Purging filter hasn't fully completed yet")
+        return False
+
+    return True
+
+
+def _purge_filtered_states(
+    session: Session,
+    excluded_entity_ids: List[str],
+    batch_purge_before: datetime,
+    batch_size_hours: int,
+) -> datetime:
+    """Handle purging filtered states."""
+
+    query = (
+        session.query(States)
+        .filter(States.entity_id.in_(excluded_entity_ids))
+        .order_by(States.last_updated.asc())
+        .limit(1)
+    )
+    states = execute(query, to_native=True, validate_entity_ids=False)
+    if states:
+        batch_purge_before = states[0].last_updated + timedelta(hours=batch_size_hours)
+
+    _LOGGER.debug(
+        "Purging states before %s that should be filtered",
+        batch_purge_before.strftime(DEBUG_DATETIME_FMT),
+    )
+
+    disconnected_rows = (
+        session.query(States)
+        .filter(
+            States.old_state_id.in_(
+                session.query(States.state_id)
+                .filter(States.last_updated < batch_purge_before)
+                .filter(States.entity_id.in_(excluded_entity_ids))
+                .subquery()
+            )
+        )
+        .update({"old_state_id": None}, synchronize_session=False)
+    )
+    _LOGGER.debug("Updated %s states to remove old_state_id", disconnected_rows)
+
+    event_ids: List[int] = [
+        event_id
+        for (event_id,) in session.query(States.event_id)
+        .filter(States.last_updated < batch_purge_before)
+        .filter(States.event_id is not None)
```

Comment:
This shouldn't be possible.  Did you find a row like this?

```suggestion
        .filter(States.event_id is not None)
```

Reply:
No I haven't. That was one case where I wasn't sure if it is implied. Technically it would be possible since the schema does allow it (`event_id` doesn't have `nullable=false`).
Personally I would like to leave it in if that's alright with you. Performance wise this shouldn't matter.